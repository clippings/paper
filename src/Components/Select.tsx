import React, { useState } from 'react';
import classnames from 'classnames';
import {
  createFormElementClassName,
  createFormElementClassNameModifier,
} from '../utils/classNames';
import '../assets/scss/select.scss';
import { Icon, ICONS, ICONS_COLORS, ICONS_DIRECTIONS } from './Icon';
const classNames = {
  wrapper: createFormElementClassName('select'),
  labelWrap: createFormElementClassNameModifier('select', 'label-wrap'),
  label: createFormElementClassNameModifier('select', 'label'),
  dropdown: createFormElementClassNameModifier('select', 'dropdown'),
  dropdownItem: createFormElementClassNameModifier('select', 'dropdown__item'),
  selected: createFormElementClassNameModifier('select', 'dropdown__item-selected'),
};
type SelectProps = {
  options: {
    value?: any | number | string;
    label?: string | React.ReactNode;
  }[];
  selected?: any | number | string;
  placeholder?: string;
  onChange: (...args: any[]) => any;
};
export const Select: React.FunctionComponent<SelectProps> = ({
  options,
  selected,
  placeholder,
  onChange,
}) => {
  const selectedOptions = options.filter(option => option.value === selected);
  const [selectedOption, setSelectedOption] = useState(
    selectedOptions.length > 0 ? selectedOptions[0] : null
  );
  const [isOpen, setIsOpen] = useState(false);
  const onValueChange = option => {
    setSelectedOption(option);
    onChange(option.value);
    setIsOpen(false);
  };
  return (
    <div
      className={classnames(classNames.wrapper, {
        open: isOpen,
      })}
    >
      <div className={classNames.labelWrap} role="presentation" onClick={() => setIsOpen(!isOpen)}>
        <span className={classNames.label}>
          {selectedOption !== null ? selectedOption.label : placeholder}
        </span>
        <Icon icon={ICONS.CHEVRON} direction={ICONS_DIRECTIONS.DOWN} color={ICONS_COLORS.DARK} />
      </div>
      <ul className={classNames.dropdown}>
        {options.map(option => {
          const selectedClass = option === selectedOption ? classNames.selected : '';
          return (
            <li
              role="presentation"
              className={classnames(classNames.dropdownItem, selectedClass)}
              onClick={() => onValueChange(option)}
              key={option.value}
            >
              {option.label}
            </li>
          );
        })}
      </ul>
    </div>
  );
};
Select.defaultProps = {
  placeholder: '',
  selected: null,
};
