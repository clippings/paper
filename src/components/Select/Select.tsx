import React, { useRef, useState } from 'react';
import classnames from 'classnames';
import '@paper/assets/scss/select.scss';
import { ICON, ICONS_COLOR } from '@paper/enums';
import { Icon } from '@paper/components';
import { useDropdown } from '@paper/utils';
import classNames from '@core/config/ClassNames';
import { SelectPropsType } from './types/SelectPropsType';

export const Select: React.FunctionComponent<SelectPropsType> = ({
  options,
  selected = null,
  placeholder = '',
  onChange,
}) => {
  const selectedOptions = options.filter(option => option.value === selected);

  const [selectedOption, setSelectedOption] = useState(
    selectedOptions.length > 0 ? selectedOptions[0] : null
  );

  const ref = useRef<HTMLDivElement | null>(null);
  const { isOpen, handleToggle, handleClose } = useDropdown(ref);

  const onValueChange = option => {
    setSelectedOption(option);
    onChange(option.value);
    handleClose();
  };

  return (
    <div
      className={classnames(classNames.select.wrapper, {
        open: isOpen,
      })}
      ref={ref}
    >
      <div className={classNames.select.labelWrap} role="presentation" onClick={handleToggle}>
        <span
          className={
            selectedOption === null ? classNames.select.labelDefault : classNames.select.label
          }
        >
          {selectedOption !== null ? selectedOption.label : placeholder}
        </span>
        <Icon
          icon={ICON.CHEVRON_DOWN}
          color={ICONS_COLOR.DARK}
          className={classNames.select.icon}
        />
      </div>
      <ul className={classNames.select.dropdown}>
        {options.map(option => {
          const selectedClass = option === selectedOption ? classNames.select.selected : '';
          return (
            <li
              role="presentation"
              className={classnames(classNames.select.dropdownItem, selectedClass)}
              onClick={() => onValueChange(option)}
              key={option.value}
            >
              {option.label}
            </li>
          );
        })}
      </ul>
    </div>
  );
};
