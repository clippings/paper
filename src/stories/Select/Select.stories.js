import React from 'react';
import figmaDecorator from 'storybook-addon-figma';
import { storiesOf } from '@storybook/react';
import { withKnobs, text } from '@storybook/addon-knobs';
import { jsxDecorator } from 'storybook-addon-jsx';
import { Select } from'../../';

const docs = require('./description.md');

const options = [
  {
    label: 'Test',
    value: 'value',
  },
];

const selectInput = () => {
  const placeholder = text('Placeholder', 'Select a value');
  const onChange = value => {
    alert(value);
  };

  return <Select placeholder={placeholder} onChange={onChange} options={options} />;
};

const figmaFile = figmaDecorator({
  url:
    'https://www.figma.com/file/2wb8Kz7wTsJmSiPMyRy8Tc/Build-%E2%80%94%C2%A0Materials-library?node-id=505%3A11649',
});

storiesOf('Paper/Inputs', module)
  .add('Select input', selectInput, { notes: { markdown: docs } })
  .addDecorator(jsxDecorator)
  .addDecorator(withKnobs)
  .addDecorator(figmaFile);
