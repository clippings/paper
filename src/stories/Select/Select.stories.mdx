import { Meta, Story, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import { Select } from '@paper/components';

<Meta title="Input/Select"
    component={Select}
/>


# Select
Basic Select component usage .

<Canvas>
  <Story
    name="Select"
    height="200px"
    args={{
      className: 'p-f-select',
      label: 'Select',
      disabled: false,
      error: '',
      options: [
        {
          label: 'Test',
          value: 'value',
        },
        {
          label: 'Test2',
          value: 'value2',
        },
      ]
    }}
  >
    {(args) => <Select {...args}/>}
  </Story>
</Canvas>

<ArgsTable story="Select"/>
