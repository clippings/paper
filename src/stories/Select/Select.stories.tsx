import React from 'react';
import figmaDecorator from 'storybook-addon-figma';
import { storiesOf } from '@storybook/react';
import { withKnobs, text } from '@storybook/addon-knobs';
import { Select } from '@paper';

import docs from './description.md';
import {jsxDecorator} from "storybook-addon-jsx";

const options = [
  {
    label: 'Test',
    value: 'value',
  },
  {
    label: 'Test2',
    value: 'value2',
  },
];

const selectInput = () => {
  const placeholder = text('Placeholder', 'Select a value');
  const onChange = value => {
    alert(value);
  };

  return <Select placeholder={placeholder} onChange={onChange} options={options} />;
};

const figmaFile = figmaDecorator({
  url:
    'https://www.figma.com/file/2wb8Kz7wTsJmSiPMyRy8Tc/Build-%E2%80%94%C2%A0Materials-library?node-id=505%3A11649',
});

storiesOf('Inputs', module)
  .addDecorator(jsxDecorator)
  .addDecorator(withKnobs)
  .addDecorator(figmaFile)
  .add('Select input', selectInput, { notes: { markdown: docs } });
