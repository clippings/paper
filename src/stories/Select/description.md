## Properties

<table>
<thead>
<tr>
<th>property</th>
<th>type</th>
<th>default</th>
<th>accepted values</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>options</strong></td>
<td>array<{label: string | ReactNode, value: string | object | number}></td>
<td></td>
<td></td>
<td>
    The lists of options to be disapleyd by the component
</td>
</tr>
<tr>
<td><strong>placeholder</strong></td>
<td>string</td>
<td></td>
<td></td>
<td>
  Used as a placeholder when no option is selected.
</td>
</tr>
<tr>
<td><strong>onChange</strong></td>
<td>function</td>
<td></td>
<td></td>
<td>
   The callback to fetch the value. Recives the option value as an argument.
</td>
</tr>
<tr>
<td><strong>selected</strong></td>
<td>string | object | number</td>
<td>null</td>
<td></td>
<td>The selected by default option</td>
</tr>
</tbody>
</table>
