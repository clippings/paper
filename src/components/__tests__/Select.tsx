import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Select } from '../Select/Select';
const onChange = jest.fn();

const options = [
  {
    label: 'Test',
    value: 1,
  },
];

const select = <Select options={options} onChange={onChange} />;

describe('Select component', () => {
  test('matches snapshot', () => {
    const { container } = render(select);
    expect(container).toMatchSnapshot();
  });

  test('onChange is triggered', () => {
    const { getByText } = render(select);
    const option = getByText('Test');

    fireEvent.click(option);

    expect(onChange).toBeCalled();
    expect(onChange).toBeCalledWith(1);
  });
});
