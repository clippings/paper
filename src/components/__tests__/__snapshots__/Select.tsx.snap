// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component matches snapshot 1`] = `
<div>
  <div
    class="p-f-select"
  >
    <div
      class="p-f-select__label-wrap"
      role="presentation"
    >
      <span
        class="p-f-select__label-default"
      />
      <svg
        class="p-f-select__icon"
        fill="#141414"
        height="16"
        preserveAspectRatio="xMidYMid meet"
        viewBox="0 0 14 14"
        width="16"
      >
        <path
          clip-rule="evenodd"
          d="M7 11.3L1.46968 5.76967L2.53033 4.70901L7 9.17868L11.4697 4.70901L12.5303 5.76967L7 11.3Z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <input
      name=""
      type="hidden"
      value=""
    />
    <ul
      class="p-f-select__dropdown"
    >
      <li
        class="p-f-select__dropdown__item"
        role="presentation"
      >
        Test
      </li>
    </ul>
  </div>
</div>
`;
